### PR TITLE
Add signature verification

### PR DESCRIPTION
**Pull request by corycaywood**
# Description
This adds public key fetching and caching as well as verification of the signature of the config after fetching and after retrieving of the the cached config from file.

An improvement that I still need to work on is implementing safeguards against modifying the cached public key in shared preferences at `PublicKeyCache` (it can be modified on a rooted device). A solution for this might be to encrypt the public key using the Android Keystore before caching it.

## Links
SDKCF-1094

# Checklist
- [x] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [x] I wrote/updated tests for new/changed code
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

File: remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/api/PublicKeyFetcher.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+internal class PublicKeyFetcher(private val client: ConfigApiClient) {
+
+    fun fetch(keyId: String): String {
+        val response = client.fetchPath("keys/$keyId")
+
+        if (!response.isSuccessful) {
+            throw IOException("Unexpected response when fetching public key: $response")
+        }
+
+        val body = response.body()!!.string()
```
Comment (donnie-jp) [inline remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/api/PublicKeyFetcher.kt]: Is this force-unwrapping safe here?

File: remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/ConfigVerifier.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            val publicKey = cache[config.keyId] ?: return false
+
+            val isVerified = signatureVerifier.verify(
+                publicKey,
+                config.values.toInputStream(),
+                config.signature
+            )
+
+            if (isVerified) {
+                true
```
Comment (donnie-jp) [inline remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/ConfigVerifier.kt:27]: The return is implicit from the return before the try/catch?

File: remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/SignatureVerifier.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            .generatePublic(keySpec) as ECPublicKey
+    }
+
+    private fun ecParameterSpecForCurve(curveName: String): ECParameterSpec {
+        val kpg = KeyPairGenerator.getInstance("EC")
+        kpg.initialize(ECGenParameterSpec(curveName))
+
+        return (kpg.generateKeyPair().public as ECPublicKey).params
+    }
+
```
Comment (donnie-jp) [inline remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/SignatureVerifier.kt:59]: If this all works I'm just saying LGTM!

**Review by donnie-jp — APPROVED** [review]

**Comment by donnie-jp** [comment]
I left a couple of comments but looks great!

Also, the ci job failed.

File: remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/api/PublicKeyFetcher.kt (hunk 1 above)
Comment (corycaywood) [reply remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/api/PublicKeyFetcher.kt]: Yeah, if `response.isSuccessful()` is true, then the body should never be null. I should put a comment here about that.

File: remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/ConfigVerifier.kt (hunk 2 above)
Comment (corycaywood) [reply remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/ConfigVerifier.kt:27]: Yep, try/catch can be used as an expression.

File: remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/SignatureVerifier.kt (hunk 3 above)
Comment (corycaywood) [reply remote-config/src/main/kotlin/com/rakuten/tech/mobile/remoteconfig/verification/SignatureVerifier.kt:59]: Yeah, this seems a little weird, but it's the only thing I could find that works on all of our supported versions. Normally you can use the `AlgorithmParameters` class to get the params, but `EC` is only supported by Android 26+ :(

**Pull request merged**